### PR TITLE
[Datasets] Fix map_groups to work with different output type

### LINKED ISSUE #29175: [data] map_groups doesn't allow returning a list of simple values
It seems like unlike map_batches, with map_groups you have to return a pd.DataFrame, even if you want to return a list of simple results. I would expect returning a list to create a dataset of simple list blocks:

```
import ray
import pandas as pd

ds = ray.data.from_items([
    {"group": 1, "value": 1},
    {"group": 1, "value": 2},
    {"group": 2, "value": 3},
    {"group": 2, "value": 4},
])

def func(group):
    return [1,2,3]

ds.groupby("group").map_groups(func).show()
```

```
(_map_block_nosplit pid=657829) Traceback (most recent call last):
(_map_block_nosplit pid=657829)   File "python/ray/_raylet.pyx", line 662, in ray._raylet.execute_task
(_map_block_nosplit pid=657829)   File "python/ray/_raylet.pyx", line 666, in ray._raylet.execute_task
(_map_block_nosplit pid=657829)   File "/home/eric/.local/lib/python3.8/site-packages/ray/data/_internal/compute.py", line 449, in _map_block_nosplit
(_map_block_nosplit pid=657829)     for new_block in block_fn(block, *fn_args, **fn_kwargs):
(_map_block_nosplit pid=657829)   File "/home/eric/.local/lib/python3.8/site-packages/ray/data/dataset.py", line 457, in transform
(_map_block_nosplit pid=657829)     applied = batch_fn(view, *fn_args, **fn_kwargs)
(_map_block_nosplit pid=657829)   File "/home/eric/.local/lib/python3.8/site-packages/ray/data/grouped_dataset.py", line 282, in group_fn
(_map_block_nosplit pid=657829)     builder.add_block(applied)
(_map_block_nosplit pid=657829)   File "/home/eric/.local/lib/python3.8/site-packages/ray/data/_internal/table_block.py", line 57, in add_block
(_map_block_nosplit pid=657829)     raise TypeError(
(_map_block_nosplit pid=657829) TypeError: Got a block of type <class 'list'>, expected <class 'pandas.core.frame.DataFrame'>.If you are mapping a function, ensure it returns an object with the expected type. Block:
```

### PR DESCRIPTION
**Pull request by c21**
Signed-off-by: Cheng Su <scnju13@gmail.com>

## Why are these changes needed?

This is to fix https://github.com/ray-project/ray/issues/29175 where `map_groups` does not work with different output type. We should use `DelegatingBlockBuilder` to create block builder based on actual block type, but not based on input block type.
 
## Related issue number

Closes https://github.com/ray-project/ray/issues/29175

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by jianoaix — APPROVED** [review]

**Review by ericl — APPROVED** [review]

File: python/ray/data/grouped_dataset.py [hunk 1]
```diff
@@ -317,7 +318,7 @@ def group_fn(batch):
                 boundaries = get_key_boundaries(block_accessor)
             else:
                 boundaries = [block_accessor.num_rows()]
-            builder = block_accessor.builder()
+            builder = DelegatingBlockBuilder()
```
Comment (ericl) [inline python/ray/data/grouped_dataset.py:321]: Could we also update the docstring of the map_groups to include this as an example?

Comment (c21) [reply python/ray/data/grouped_dataset.py:330]: @ericl - sure, added one. Please check if the example makes sense - get first value per group.

**Review by ericl — COMMENTED** [review]
Example LGTM

File: python/ray/data/grouped_dataset.py [hunk 2]
```diff
@@ -247,14 +248,23 @@ def map_groups(
...
             >>> import ray
             >>> import pandas as pd
             >>> import numpy as np
+            >>> # Get median per group. Note that median is not an associative
+            >>> # function so cannot be computed with aggregate().
             >>> ds = ray.data.range(100) # doctest: +SKIP
             >>> ds.groupby(lambda x: x % 3).map_groups( # doctest: +SKIP
             ...     lambda x: [np.median(x)])
+            >>> # Get first value per group.
+            >>> ds = ray.data.from_items([ # doctest: +SKIP
```
Comment (jianoaix) [inline python/ray/data/grouped_dataset.py:261]: Can those SKIPs be removed? It looks they should be able to run without issue in case here.

Comment (c21) [reply python/ray/data/grouped_dataset.py:261]: @jianoaix - I guess they can be removed. Just to follow what we have here to be safe. Wanted to get this in before branch cut as well. I can audit all SKIPs in file later as followup.

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/tests/test_dataset.py [hunk 3]
```diff
@@ -4332,10 +4332,6 @@ def test_groupby_map_groups_merging_invalid_result(ray_start_regular_shared):
     with pytest.raises(TypeError):
         grouped.map_groups(lambda x: None if x == [1] else x)
 
-    # The UDF returns a type that's different than the input type, which is invalid.
-    with pytest.raises(TypeError):
-        grouped.map_groups(lambda x: pd.DataFrame([1]) if x == [1] else x)
```
Comment (c21) [inline python/ray/data/tests/test_dataset.py:4337]: This failed test does not make sense after this PR. Removing.

**Comment by c21** [comment]
`buildkite/oss-ci-build-pr/book-documentation` and Datasets CI test are passed.

**Pull request merged**